File: src/algorithms/array-techniques/sliding-window/sliding-window.ts
export interface SlidingWindowResult {
  maxSum: number;
  windowStartIndex: number;
}

export function maxSumSubarray(inputArray: number[], windowSize: number): SlidingWindowResult {
  if (inputArray.length === 0 || windowSize <= 0 || windowSize > inputArray.length) {
    return { maxSum: 0, windowStartIndex: 0 };
  }

  /* Compute the sum of the initial window */
  let currentSum = 0;
  for (let initIndex = 0; initIndex < windowSize; initIndex++) {
    currentSum += inputArray[initIndex]!;
  }

  let maxSum = currentSum;
  let windowStartIndex = 0;

  /* Slide the window to the right */
  for (let rightIndex = windowSize; rightIndex < inputArray.length; rightIndex++) {
    currentSum -= inputArray[rightIndex - windowSize]!;
    currentSum += inputArray[rightIndex]!;

    if (currentSum > maxSum) {
      maxSum = currentSum;
      windowStartIndex = rightIndex - windowSize + 1;
    }
  }

  return { maxSum, windowStartIndex };
}

File: src/algorithms/array-techniques/sliding-window/sources/sliding-window.py
def max_sum_subarray(input_array: list[int], window_size: int) -> dict:
    if len(input_array) == 0 or window_size <= 0 or window_size > len(input_array):
        return {"max_sum": 0, "window_start_index": 0}

    current_sum = 0
    for init_index in range(window_size):
        current_sum += input_array[init_index]

    max_sum = current_sum
    window_start_index = 0

    for right_index in range(window_size, len(input_array)):
        current_sum -= input_array[right_index - window_size]
        current_sum += input_array[right_index]

        if current_sum > max_sum:
            max_sum = current_sum
            window_start_index = right_index - window_size + 1

    return {"max_sum": max_sum, "window_start_index": window_start_index}
